### PR TITLE
Refactor: UTXO probing in TransactionPool

### PR DESCRIPTION
**Pull request by omerfirmak**
Existing implementation was buggy, had paths that could SEGV us.

### CODE REVIEW

File: source/agora/node/FullNode.d [hunk 1]
```diff
@@ -1005,66 +1005,24 @@ public class FullNode : API
...
-            {
-                foreach (ref Hash hash, ref Transaction tx; this.pool)
-                {
-                    if (!tx.isFreeze())
-                        continue;
-
-                    foreach (size_t idx, output_; tx.outputs)
-                    {
-                        if (output_.type == OutputType.Freeze &&
-                            output_.value >= Amount.MinFreezeAmount)
```
Comment (mkykadir) [inline source/agora/node/FullNode.d:1056]: What about these checks?

Comment (omerfirmak) [reply source/agora/node/FullNode.d:1056]: enrollment validation routine checks that anyways, so they were redundant

**Review by mkykadir — APPROVED** [review]

**Review by linked0 — APPROVED** [review]
LGTM. Thanks.

File: source/agora/consensus/pool/Transaction.d [hunk 2]
```diff
@@ -570,6 +580,29 @@ public class TransactionPool
...
+
+    public bool peekUTXO (in Hash utxo, out UTXO value) nothrow @safe
+    {
+        if (auto output = utxo in this.utxo_set)
+        {
+            value.output = *output;
+            return true;
+        }
+        return false;
+    }
```
Comment (Geod24) [inline source/agora/consensus/pool/Transaction.d:605]: Looks to me like we should just have another `UTXOCache` here ?

Comment (omerfirmak) [reply source/agora/consensus/pool/Transaction.d:605]: That was my initial thought too, but to remove a UTXO we would either need to construct a dummy spending TX or expose the private `remove` UTXOCache has. Either way didnt feel right, so this.

I can change it to use UTXOCache if you are comfortable with one of those.

Comment (Geod24) [reply source/agora/consensus/pool/Transaction.d:605]: That answers my question

**Pull request merged**